### PR TITLE
Fixes for in-tree scripts: offline mode, maybe-build for fqltool-test…

### PR DESCRIPTION
**Pull request by michaelsembwever**
…, jvm-dtest-upgrade

 - jvm-dtest-upgrade script running in ci-cassandra.a.o (needs default bash)
 - add offline mode (to avoid `git fetch` before building dtest-jars on other branches),
 - avoid rebuilding in fqltool-test target, as is standard for test targets.
 - python detection and override
 - FileTest cleanup (tmp directory was left unreadable and undeletable)
 - reuse ~/.ccm (like normal)
 - remove support for python 3.6

CASSANDRA-18665

### CODE REVIEW

**Comment by michaelsembwever** [comment]
@Mmuzaf

File: .build/run-python-dtests.sh [hunk 1]
```diff
@@ -73,7 +72,7 @@ if [ "${java_version}" -eq 17 ] && [[ "${target}" == "dtest-upgrade" ]] ; then
     exit 1
 fi
 
-python_version=$(python -V | awk '{print $2}' | awk -F'.' '{print $1"."$2}')
+python_version=$(python -V 2>&1 | awk '{print $2}' | awk -F'.' '{print $1"."$2}')
```
Comment (Mmuzaf) [inline .build/run-python-dtests.sh:75]: Since we removed the 3.6 version, should we add a constraint here that version >= 3.7?

**Review by Mmuzaf — COMMENTED** [review]
.build/run-python-dtests.sh:

`sed -r "s/<[\/]?testsuites>//g" ${DIST_DIR}/test/output/nosetests.xml > /tmp/nosetests.xml`

Should we save an intermediate result to the `${TMPDIR}` ?

File: .build/run-python-dtests.sh (hunk 1 above)
Comment (michaelsembwever) [reply .build/run-python-dtests.sh:75]: in bash how would you do that ?

Comment (Mmuzaf) [reply .build/run-python-dtests.sh:75]: `cassandra_regx_supported_versions="^(2.2|3.0|3.11|4.0|4.1)(.[0-9]+)?$"`

Like this one in the accompanying script in the [cassandra-builds](https://github.com/apache/cassandra-builds/pull/96/files#diff-f458d745bb6227c4f9161c75ffed2a98d0ee3cd7da734afe7e759670c5969b3cR23) :-)

Comment (michaelsembwever) [reply .build/run-python-dtests.sh:75]: done in https://github.com/apache/cassandra/compare/90182972ebc0dee7f936058f222d2df966893d1d..557f90ed5ccdaa1e86459e135017ef96cb64bc4b

**Comment by michaelsembwever** [comment]
> Should we save an intermediate result to the ${TMPDIR} ? 

done in https://github.com/apache/cassandra/compare/90182972ebc0dee7f936058f222d2df966893d1d..557f90ed5ccdaa1e86459e135017ef96cb64bc4b

File: .build/docker/ubuntu2004_test.docker [hunk 2]
```diff
@@ -146,11 +131,26 @@ RUN /bin/bash -c "export CASS_DRIVER_NO_CYTHON=1 CASS_DRIVER_NO_EXTENSIONS=1 \
...
-    ccm create -n 1 -v git:trunk test && ccm remove test && \
+RUN /bin/bash -c "source ${BUILD_HOME}/env3.7/bin/activate && \
     ccm create -n 1 -v git:cassandra-4.1 test && ccm remove test && \
     ccm create -n 1 -v git:cassandra-4.0 test && ccm remove test"
 
+# Initialize ccm versions.  right side of each sequence needs to be updated with new releases.
+#  this can be checked with:
+#  `curl -s https://downloads.apache.org/cassandra/ | grep -oP '(?<=href=\")[0-9]+\.[0-9]+\.[0-9]+(?=)' | sort -V | uniq -w 3`
+RUN bash -c 'source ~/env3.7/bin/activate && \
+    for i in {1..11} ; do echo $i ; ccm create --quiet -n 1 -v binary:4.0.$i test && ccm remove test ; done && \
```
Comment (Mmuzaf) [inline .build/docker/ubuntu2004_test.docker:142]: I'm not sure if it's worth doing the way below, but I think we can get 4.x releases from the `downloads'.  WDYT?

```
m@MBP ~ % curl --silent https://downloads.apache.org/cassandra/ | grep -o '\[DIR\]\"> <a href=".*">' | sed 's/\[DIR\]\"> <a href="//;s/\/">//' | grep -E '^([4-4]|[0-9]{2,})\..*$' | 

The output:

4.0.11
4.1.3
```

```
m@MBP ~ % curl --silent https://downloads.apache.org/cassandra/ | grep -o '\[DIR\]\"> <a href=".*">' | sed 's/\[DIR\]\"> <a href="//;s/\/">//' | grep -E '^([4-4]|[0-9]{2,})\..*$' | awk -F. '{ for(i=1; i<=$NF; i++) system("echo Executing command for version "$0", iteration " i) }'

Executing command for version 4.0.11, iteration 1
Executing command for version 4.0.11, iteration 2
Executing command for version 4.0.11, iteration 3
Executing command for version 4.0.11, iteration 4
Executing command for version 4.0.11, iteration 5
Executing command for version 4.0.11, iteration 6
Executing command for version 4.0.11, iteration 7
Executing command for version 4.0.11, iteration 8
Executing command for version 4.0.11, iteration 9
Executing command for version 4.0.11, iteration 10
Executing command for version 4.0.11, iteration 11
Executing command for version 4.1.3, iteration 1
Executing command for version 4.1.3, iteration 2
Executing command for version 4.1.3, iteration 3
```

Comment (michaelsembwever) [reply .build/docker/ubuntu2004_test.docker:142]: Can you get it into this line:
```
RUN bash -c 'source ~/env3.7/bin/activate && \
    for i in {1..11} ; do echo $i ; ccm create --quiet -n 1 -v binary:4.0.$i test && ccm remove test ; done && \
    for i in {1..3}  ; do echo $i ; ccm create --quiet -n 1 -v binary:4.1.$i test && ccm remove test ; done'
```

i've failed at every attempt so far…

Comment (Mmuzaf) [reply .build/docker/ubuntu2004_test.docker:142]: I've managed to improve the latest version :-)

```
mmuzaf@Maxims-MBP ~ % curl --silent https://downloads.apache.org/cassandra/ | grep -o '\[DIR\]\"> <a href=".*">' | sed 's/\[DIR\]\"> <a href="//;s/\/">//' | grep -E '^([4-4]|[0-9]{2,})\..*$' | awk -F. '{ for(i=1; i<=$NF; i++) print $1 "." $2 "." i }' 
4.0.1
4.0.2
4.0.3
4.0.4
4.0.5
4.0.6
4.0.7
4.0.8
4.0.9
4.0.10
4.0.11
4.1.1
4.1.2
4.1.3
```

Comment (michaelsembwever) [reply .build/docker/ubuntu2004_test.docker:142]: it's the `RUN bash -c ' … '` that's the biggest hurdle

Comment (Mmuzaf) [reply .build/docker/ubuntu2004_test.docker:142]: Yeah, it seems this is a problem...
What if we create a script`_init_ccm.sh 4.0 4.1` and run it instead?

Comment (michaelsembwever) [reply .build/docker/ubuntu2004_test.docker:142]: separate ticket for follow up work ?

Comment (Mmuzaf) [reply .build/docker/ubuntu2004_test.docker:142]: +1 
:-)

**Review by Mmuzaf — APPROVED** [review]
I'm a little worried about the way we configure ccm, but I've seen that it's not a new code and it's actually been done before this. So, let's create an issue for that and continue with everything we have now. 

The rest of the changes are LGTM

File: .build/docker/run-tests.sh [hunk 3]
```diff
@@ -69,8 +69,11 @@ if [[ ! "${java_version}" =~ $regx_java_version ]]; then
     exit 1
 fi
 
-python_version="3.6"
-command -v python >/dev/null 2>&1 && python_version="$(python -V | awk '{print $2}' | awk -F'.' '{print $1"."$2}')"
+# allow python version override, otherwise default to current python version or 3.7
+if [ "x" == "x${python_version}" ] ; then
+    command -v python >/dev/null 2>&1 && python_version="$(python -V 2>&1 | awk '{print $2}' | awk -F'.' '{print $1"."$2}')"
+    python_version="${python_version:-3.7}"
```
Comment (jmckenzie-dev) [inline .build/docker/run-tests.sh:75]: I don't have a horse in this race, but why 3.7? vs. 3.8, 3.11, etc.

File: .build/docker/ubuntu2004_test.docker [hunk 4]
```diff
@@ -41,7 +38,7 @@ RUN echo 'Acquire::http::Timeout "60";' > /etc/apt/apt.conf.d/80proxy.conf
 RUN echo 'Acquire::ftp::Timeout "60";' >> /etc/apt/apt.conf.d/80proxy.conf
 
 # install our python dependencies and some other stuff we need
-# libev4 libev-dev are for the python driver / libssl-dev is for python3.6
+# libev4 libev-dev are for the python driver
```
Comment (jmckenzie-dev) [inline .build/docker/ubuntu2004_test.docker:41]: Are we dropping python3.6? Asking for a friend. 😀

File: .build/docker/ubuntu2004_test.docker (hunk 2 above)
Comment (jmckenzie-dev) [inline .build/docker/ubuntu2004_test.docker:142]: So we're intentionally trying to init 4.0.1 to 4.0.11 right? And when a 4.0.12 hits we'll need to manually add that here?

If we want to be... clever? Hacky? We could do something like:
`git tag | grep -E 'cassandra-4.0.[0-9]+$' | cut -d '.' -f 3 | sort -nr | head -n 1`

to parse out the latest version at time of docker image creation and iterate through all up to that point to init them. Alternatively, or along with that, might be nice to add a comment to indicate why we have the top end of the sequence here since it's non-obvious.

File: .build/run-tests.sh [hunk 5]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/bin/bash
```
Comment (jmckenzie-dev) [inline .build/run-tests.sh:1]: Was this an accidental change or intentional? If intentional (for my own education) - why the change?

File: .build/run-tests.sh [hunk 6]
```diff
@@ -84,7 +87,9 @@ _build_all_dtest_jars() {
     fi
 
     if [ -d ${TMP_DIR}/cassandra-dtest-jars ] && [ "https://github.com/apache/cassandra.git" == "$(git -C ${TMP_DIR}/cassandra-dtest-jars remote get-url origin)" ] ; then
-        until git -C ${TMP_DIR}/cassandra-dtest-jars fetch --quiet origin ; do echo "git pull failed… trying again… " ; done
+      if [ "x" == "x${OFFLINE}" ] ; then
+        until git -C ${TMP_DIR}/cassandra-dtest-jars fetch --quiet origin ; do echo "git fetch failed… trying again… " ; done
```
Comment (jmckenzie-dev) [inline .build/run-tests.sh:91]: Can this infinite loop / hang? And if so maybe we constrain it and explain why the need?

**Review by jmckenzie-dev — COMMENTED** [review]
Think it's fine to merge in its current form though there's a couple/few thing that might be nice to comment or clarify on commit.

**Review by jmckenzie-dev — APPROVED** [review]
... I hit the wrong radio button. Meant to approve. ;)

File: .build/docker/ubuntu2004_test.docker (hunk 2 above)
Comment (Mmuzaf) [reply .build/docker/ubuntu2004_test.docker:142]: I thought about it too, but this approach may fail the CI if the git tag is added before the actual release happens, right? That's why fetching from the archive is better, I think...

File: .build/docker/ubuntu2004_test.docker (hunk 4 above)
Comment (michaelsembwever) [reply .build/docker/ubuntu2004_test.docker:41]: yes.  it simply does not work at all anymore.

File: .build/docker/run-tests.sh (hunk 3 above)
Comment (michaelsembwever) [reply .build/docker/run-tests.sh:75]: just a conservative/minimal bump after dropping 3.6

(am all for discussing removing 3.7 too (in a separate ticket))

File: .build/run-tests.sh (hunk 6 above)
Comment (michaelsembwever) [reply .build/run-tests.sh:91]: it can, but you get the echo statements so it's visible to the user. they can ctl-c it.

this poormans approach is repeated in a number of places.  improving it (which is smart) should be done together.  and maybe just rewriting bash altogether into something non-shit

File: .build/run-tests.sh (hunk 5 above)
Comment (michaelsembwever) [reply .build/run-tests.sh:1]: I saw it break in some environments (where it picked up custom bash command on the path).
this was the simple fix. i haven't yet heard of an env where /bin/bash does not exist.

this approach is also the standard for the other scripts.

**Comment by michaelsembwever** [comment]
Committed as https://github.com/apache/cassandra/commit/987d03c142f086f05857f9c6a2eafef7e157eb15
